File: scrape/new.py
"""
Fetch new products.
Fetch detailed information about products.

CRON JOB: 0 05 1,14 * *
"""

import os
import time
from typing import List
import concurrent.futures

import pymongo
import requests
from pymongo.mongo_client import MongoClient

from proxy import Proxy


_CLIENT = MongoClient(
    f'mongodb+srv://{os.environ.get("mongodb_username")}:{os.environ.get("mongodb_password")}'
    f'@vinskraper.wykjrgz.mongodb.net/'
    f'?retryWrites=true&w=majority&appName=vinskraper'
)
_DATABASE = _CLIENT['vinskraper']['varer']

_PROXY = Proxy()

_NEW = "https://www.vinmonopolet.no/vmpws/v2/vmp/search?searchType=product&currentPage={}&q=%3Arelevance%3AnewProducts%3Atrue"
_DETAILS = 'https://www.vinmonopolet.no/vmpws/v3/vmp/products/{}?fields=FULL'

_MONTH = time.strftime('%Y-%m-01')
_IMAGE = {'thumbnail': 'https://bilder.vinmonopolet.no/bottle.png',
          'product': 'https://bilder.vinmonopolet.no/bottle.png'}


def _process_images(images):
    return {img['format']: img['url'] for img in images} if images else _IMAGE


def _process(products) -> List[dict]:
    return [{
        'index': int(product.get('code', 0)),
        'navn': product.get('name', None),
        'volum': product.get('volume', {}).get('value', 0.0),
        'land': product.get('main_country', {}).get('name', None),
        'distrikt': product.get('district', {}).get('name', None),
        'underdistrikt': product.get('sub_District', {}).get('name', None),
        'kategori': product.get('main_category', {}).get('name', None),
        'underkategori': product.get('main_sub_category', {}).get('name', None),
        'url': f'https://www.vinmonopolet.no{product.get("url", "")}',
        'status': product.get('status', None),
        'kan kjøpes': product.get('buyable', False),
        'utgått': product.get('expired', True),

        'produktutvalg': product.get('product_selection', None),
        'bærekraftig': product.get('sustainable', False),
        'bilde': _process_images(product.get('images')),
        f'pris {_MONTH}': product.get('price', {}).get('value', 0.0),
        'literpris': 100 * product.get('price', {}).get('value', 0.0) / product.get('volume', {}).get('value', 1.0),
        'ny': 0,

        'tilgjengelig for bestilling': product.get(
            'productAvailability', {}
        ).get(
            'deliveryAvailability', {}
        ).get(
            'availableForPurchase', False
        ),

        'bestillingsinformasjon': product.get(
            'productAvailability', {}
        ).get(
            'deliveryAvailability', {}
        ).get(
            'infos', [{}]
        )[0].get(
            'readableValue', None
        ),

        'tilgjengelig i butikk': product.get(
            'productAvailability', {}
        ).get(
            'storesAvailability', {}
        ).get(
            'availableForPurchase', False
        ),

        'butikkinformasjon': product.get(
            'productAvailability', {}
        ).get(
            'storesAvailability', {}
        ).get('infos', [{}]
              )[0].get(
            'readableValue', None
        ),
    } for product in products]


def _news(page: int) -> List[dict]:
    """Extract new products from a single page."""
    for _ in range(10):
        try:
            response = requests.get(_NEW.format(page), proxies=_PROXY.get(), timeout=3)
            if response.status_code != 200:
                raise ValueError()
            products = response.json().get('productSearchResult', {}).get('products', [])
            return _process(products)
        except Exception as err:
            print(f'Error: Trying another proxy. {err}')
            _PROXY.renew()
    else:
        raise ValueError('Failed to fetch new products. Tried 10 times.')


def _details(product: int) -> dict:
    """Extract detailed information about a single product."""
    for _ in range(10):
        try:
            details = requests.get(_DETAILS.format(product),
                                   proxies=_PROXY.get(),
                                   timeout=3)
            if details.status_code != 200:
                raise ValueError()
            details = details.json()
            return {
                'index': int(details.get('code', 0)),

                'navn': details.get('name', None),
                'volum': details.get('volume', {}).get('value', 0.0),
                'land': details.get('main_country', {}).get('name', None),
                'distrikt': details.get('district', {}).get('name', None),
                'underdistrikt': details.get('sub_District', {}).get('name', None),
                'kategori': details.get('main_category', {}).get('name', None),
                'underkategori': details.get(
                    'main_sub_category', {}
                ).get('name', None),
                'url': f'https://www.vinmonopolet.no{details.get("url", "")}',
                'status': details.get('status', None),
                'kan kjøpes': details.get('buyable', False),
                'utgått': details.get('expired', False),
                'tilgjengelig for bestilling': details.get(
                    'productAvailability', {}
                ).get(
                    'deliveryAvailability', {}
                ).get(
                    'availableForPurchase', False
                ),
                'bestillingsinformasjon': details.get(
                    'productAvailability', {}
                ).get(
                    'deliveryAvailability', {}
                ).get(
                    'infos', [{}]
                )[0].get(
                    'readableValue', None
                ),
                'produktutvalg': details.get('product_selection', None),
                'bærekraftig': details.get('sustainable', False),
                'bilde': _process_images(details.get('images')),
                f'pris {_MONTH}': details.get('price', {}).get('value', 0.0),
                'ny': 0,

                'farge': details.get('color', None),
                'karakteristikk': [characteristic['readableValue'] for characteristic in
                                   details.get('content', {}).get('characteristics', [])],
                'ingredienser': [ingredient['readableValue'] for ingredient in
                                 details.get('content', {}).get('ingredients', [])],
                **{trait["name"].lower(): trait["readableValue"]
                    for trait in details.get('content', {}).get('traits', [])},
                'lukt': details.get('smell', None),
                'smak': details.get('taste', None),
                'allergener': details.get('allergens', None),

                'passer til': [element['name'] for element in details.get('content', {}).get('isGoodFor', [])],
                'lagring': details.get('content', {}).get('storagePotential', {}).get('formattedValue', None),
                'kork': details.get('cork', None),

                'beskrivelse': {'lang': details.get('content', {}).get('style', {}).get('description', None),
                                'kort': details.get('content', {}).get('style', {}).get('name', None)},
                'metode': details.get('method', None),
                'årgang': details.get('year', None),

                'literpris': details.get('litrePrice', {}).get('value', 0.0),
            }
        except Exception as err:
            print(f'{product}: Trying another proxy. {err}')
            _PROXY.renew()

    raise ValueError('Failed to fetch product information.')


def details(products: List[int] = None, max_workers=5):
    """
    Fetch detailed information about products and store the results to the database.

    Parameters
    ----------
    products : List[int]
        The products to fetch detailed information about.
        If None, all products are fetched.
    max_workers : int
        The maximum number of (parallel) workers to use when fetching products.

    Raises
    ------
    ValueError
        If the product information could not be fetched.

    Notes
    -----
    The function fetches detailed information about the products in the list `products`.
    If `products` is None, all products are fetched.
    The function uses a thread pool to fetch the products in parallel.
    To prevent memory issues, the function fetches the products in batches.
    """
    if not products:
        print('Fetching all products.')
        products = _DATABASE.distinct('index')

    step = max(len(products) // 1000, 500)
    for i in range(0, len(products), step):
        print(f'Processing products {i} to {i + step} of {len(products)}.')

        operations = []
        with concurrent.futures.ThreadPoolExecutor(max_workers=max_workers) as executor:
            results = {executor.submit(_details, product): product for product in products[i:i + step]}
            for future in concurrent.futures.as_completed(results):
                product = future.result()
                if not product:
                    continue

                if 'alkohol' in product:
                    try:
                        product['alkohol'] = float(product['alkohol'].replace(' prosent', '').replace(',', '.'))
                    except Exception:
                        product['alkohol'] = 0.0
                else:
                    product['alkohol'] = 0.0
                if product['årgang']:
                    product['årgang'] = int(product['årgang'])

                product = {key: value if value else None for key, value in product.items()}

                operations.append(
                    pymongo.UpdateOne(
                        {'index': product['index']},
                        {'$set': product},
                        upsert=True
                    )
                )
        _DATABASE.bulk_write(operations)


def news(max_workers=5):
    _DATABASE.update_many(
        {},
        {'$inc': {'ny': 1}}  # Increment 'ny' field by 1 for all documents.
    )

    expired = set(_CLIENT['vinskraper']['utgått'].distinct('index'))
    response = requests.get(_NEW.format(0), proxies=_PROXY.get(), timeout=3)

    if response.status_code != 200:
        raise ValueError('Failed to fetch new products.')

    response = response.json()
    pages = response.get('contentSearchResult', {}).get('pagination', {}).get('totalPages', 0)

    ids = list(_DATABASE.find({'oppdater': True}))
    _DATABASE.update_many(
        {'oppdater': True},
        {'$unset': {'oppdater': ''},
         '$set': {'ny': 0}}
    )

    old = set(_DATABASE.distinct('index'))
    with concurrent.futures.ThreadPoolExecutor(max_workers=max_workers) as executor:
        results = list(executor.map(_news, range(pages)))

        new = list({product['index'] for result in results for product in result} - old)

        # If a new product already exists, but is expired, move it back to the main collection.
        _expired = set(new) & expired
        if _expired:
            moving = list(_CLIENT['vinskraper']['utgått'].find({'index': {'$in': list(_expired)}}))
            for record in moving:
                record['ny'] = 1
            _CLIENT['vinskraper']['utgått'].delete_many({'index': {'$in': list(_expired)}})
            _DATABASE.insert_many(moving)
            expired -= _expired

        operations = [pymongo.UpdateOne(
            {'index': product['index']},
            {'$set': product},
            upsert=True
        ) for result in results for product in result if product['index'] in new]

        result = _DATABASE.bulk_write(operations)
        print(f'Inserted {result.upserted_count} new products.')
        print(f'Updated {result.modified_count} existing products.')

        ids.extend(new)
    if ids:
        details(ids)

File: scrape/new.mjs
import axios from "axios";
import { HttpsProxyAgent } from "https-proxy-agent";
import { MongoClient, ServerApiVersion } from "mongodb";
import dotenv from "dotenv";

dotenv.config();

const client = new MongoClient(
  `mongodb+srv://${process.env.MONGO_USR}:${process.env.MONGO_PWD}@snublejuice.faktu.mongodb.net/?retryWrites=true&w=majority&appName=snublejuice`,
  {
    serverApi: {
      version: ServerApiVersion.v1,
      strict: false,
      deprecationErrors: true,
    },
  },
);
await client.connect();

const database = client.db("snublejuice");
const itemCollection = database.collection("products");
const itemIds = await itemCollection.distinct("index");

const proxies = process.env.PROXY_IPS.split(",").flatMap((ip) => [
  new HttpsProxyAgent(
    `http://${process.env.PROXY_USR}:${process.env.PROXY_PWD}@${ip}:${process.env.PROXY_PRT}`,
  ),
]);

function getNextProxy(proxy) {
  if (proxies.length === 0) {
    throw new Error("No more proxies available!");
  }

  const index = proxy ? proxies.findIndex((p) => p.proxy.host === proxy.proxy.host) : -1;
  return proxies[(index + 1) % proxies.length];
}

function removeProxy(proxy) {
  const index = proxies.findIndex((p) => p.proxy.host === proxy.proxy.host);
  if (index > -1) {
    const removedProxy = proxies.splice(index, 1)[0];
    console.log(
      `Removed failing proxy ${removedProxy.proxy.host}. ${proxies.length} proxies remaining.`,
    );
  }
}

const URL =
  "https://www.vinmonopolet.no/vmpws/v2/vmp/search?fields=FULL&searchType=product&currentPage={}&q=%3Arelevance%3AnewProducts%3Atrue";

const LINK = "https://www.vinmonopolet.no{}";

const IMAGE = {
  thumbnail: "https://bilder.vinmonopolet.no/bottle.png",
  product: "https://bilder.vinmonopolet.no/bottle.png",
};

function processImages(images) {
  return images ? images.reduce((acc, img) => ({ ...acc, [img.format]: img.url }), {}) : IMAGE;
}

function processProducts(products) {
  const processedProducts = [];

  for (const product of products) {
    const index = parseInt(product.code, 10) || null;

    if (itemIds.includes(index)) {
      break;
    }

    processedProducts.push({
      index: index,

      updated: true,

      name: product.name || null,
      price: product.price?.value || 0.0,

      volume: product.volume?.value || 0.0,
      literprice:
        product.price?.value && product.volume?.value
          ? product.price.value / (product.volume.value / 100.0)
          : 0.0,

      url: product.url ? LINK.replace("{}", product.url) : null,
      images: product.images ? processImages(product.images) : IMAGE,

      category: product.main_category?.name || null,
      subcategory: product.main_sub_category?.name || null,

      country: product.main_country?.name || null,
      district: product.district?.name || null,
      subdistrict: product.sub_District?.name || null,

      selection: product.product_selection || null,
      sustainable: product.sustainable || false,

      buyable: product.buyable || false,
      expired: product.expired || true,
      status: product.status || null,

      orderable: product.productAvailability?.deliveryAvailability?.availableForPurchase || false,
      orderinfo:
        product.productAvailability?.deliveryAvailability?.infos?.[0]?.readableValue || null,

      instores: product.productAvailability?.storesAvailability?.availableForPurchase || false,
      storeinfo: product.productAvailability?.storesAvailability?.infos?.[0]?.readableValue || null,
    });
  }

  return processedProducts;
}

async function getPage(page, proxy) {
  for (let attempt = 0; attempt < 5; attempt++) {
    try {
      const response = await session.get(URL.replace("{}", page), {
        httpsAgent: proxy,
        timeout: 10000,
      });

      if (response.status === 200) {
        return processProducts(response.data["productSearchResult"]["products"]);
      }

      console.log(`Status code ${response.status} at page ${page} (trying another proxy); ${err}`);
    } catch (err) {
      console.log(`Request failed with proxy ${proxy.host} for page ${page}: ${err.message}`);
      if (err.code === "ECONNREFUSED" || err.code === "ETIMEDOUT") {
        removeProxy(proxy);
      }
    }

    try {
      proxy = getNextProxy(proxy);
      await new Promise((resolve) => setTimeout(resolve, 2000));
    } catch (err) {
      throw new Error(`No more proxies available to fetch id ${id}`);
    }
  }

  throw new Error(`Failed to fetch page ${page} after 5 attempts.`);
}

async function updateDatabase(data) {
  return await itemCollection.insertMany(data);
}

async function getProducts() {
  let items = [];
  let proxy = getNextProxy();

  for (let page = 0; page < 10000; page++) {
    console.log(`Page ${page}.`);

    try {
      let products = await getPage(page, proxy);
      if (products.length === 0 || !products) {
        console.log(`No more new products (final page: ${page}).`);
        break;
      }

      items = items.concat(products);
      proxy = getNextProxy(proxy);

      await new Promise((resolve) => setTimeout(resolve, 1100));
    } catch (err) {
      console.log(`Error page ${page}! ${err}`);
      break;
    }

    // Upsert to the database every 10 pages.
    if (page % 10 === 0) {
      if (items.length === 0) {
        return;
      }

      console.log(`Adding ${items.length} products.`);
      const result = await updateDatabase(items);
      console.log(` Inserted ${result.insertedCount} records`);

      items = [];
    }
  }

  // Insert the remaining products, if any.
  if (items.length === 0) {
    return;
  }
  console.log(`Adding ${items.length} final products.`);
  const result = await updateDatabase(items);
  console.log(` Inserted ${result.insertedCount} records`);
}

const session = axios.create();

async function main() {
  await getProducts();
}

await main();

client.close();
